virtcontainers/acrn.go

```diff
@@ -0,0 +1,630 @@
+
+	return a.arch.memoryTopology(memMb), nil
+}
+
+func (a *acrn) appendImage(devices []Device) ([]Device, error) {
+	// Get sandbox and increment the globalIndex.
+	// This is to make sure the VM rootfs occupies
+	// the first Index which is /dev/vda.
+	sandbox, err := globalSandboxList.lookupSandbox(a.id)
+	if sandbox == nil && err != nil {
+		return nil, err
+	}
+	sandbox.GetAndSetSandboxBlockIndex()
+
+	imagePath, err := a.config.ImageAssetPath()
+	if err != nil {
+		return nil, err
+	}
+
+	if imagePath != "" {
+		devices, err = a.arch.appendImage(devices, imagePath)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return devices, nil
+}
+
+func (a *acrn) buildDevices(initrdPath string) ([]Device, error) {
+	var devices []Device
+
+	console, err := a.getSandboxConsole(a.id)
+	if err != nil {
+		return nil, err
+	}
+
+	// Add bridges before any other devices. This way we make sure that
+	// bridge gets the first available PCI address.
+	devices = a.arch.appendBridges(devices)
+
+	//Add LPC device to the list of other devices.
+	devices = a.arch.appendLPC(devices)
+
+	devices = a.arch.appendConsole(devices, console)
+
+	if initrdPath == "" {
+		devices, err = a.appendImage(devices)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	devices, err = a.createdummyVirtiBlkDev(devices)
+	if err != nil {
+		return nil, err
+	}
+
+	return devices, nil
+
+}
+
+// setup sets the Acrn structure up.
+func (a *acrn) setup(id string, hypervisorConfig *HypervisorConfig, vcStore *store.VCStore) error {
+	span, _ := a.trace("setup")
+	defer span.Finish()
+
+	err := hypervisorConfig.valid()
+	if err != nil {
+		return err
+	}
+
+	a.id = id
+	a.store = vcStore
+	a.config = *hypervisorConfig
+	a.arch = newAcrnArch(a.config)
+
+	if err = a.store.Load(store.Hypervisor, &a.state); err != nil {
+
+		// ACRN currently supports only known UUIDs for security
+		// reasons (FuSa). When launching VM, only these pre-defined
+		// UUID should be used else VM launch will fail. ACRN team is
+		// working on a solution to expose these pre-defined UUIDs
+		// to Kata in order for it to launch VMs successfully.
+		// Until this support is available, Kata is limited to
+		// launch 1 VM (using the default UUID).
+		// https://github.com/kata-containers/runtime/issues/1785
+		/*
+			a.Logger().Debug("Creating UUID")
+			a.state.UUID = uuid.Generate().String()
+		*/
+		// The path might already exist, but in case of VM templating,
+		// we have to create it since the sandbox has not created it yet.
+		if err = os.MkdirAll(store.SandboxRuntimeRootPath(id), store.DirMode); err != nil {
+			return err
+		}
+
+		if err = a.store.Store(store.Hypervisor, a.state); err != nil {
+			return err
+		}
+	}
+
+	if err = a.store.Load(store.Hypervisor, &a.info); err != nil {
+		a.Logger().WithField("function", "setup").WithError(err).Info("No info could be fetched")
+	}
+
+	return nil
+}
+
+func (a *acrn) createdummyVirtiBlkDev(devices []Device) ([]Device, error) {
```

Comment:
Can you add documentation here explaining what this function does and why we need a pool of dummy devices?

Reply:
done.